lib/meta_logger/formatter.ex

```diff
@@ -0,0 +1,98 @@
+defprotocol MetaLogger.Formatter do
+  @doc """
+  MetaLogger forrmatter protocol
+  """
+  @fallback_to_any true
+
+  @spec format(struct()) :: String.t() | list()
+  def format(payload)
+
+  defmodule IncorrectPayload do
+    defexception [:message]
+  end
+
+  defmodule IncorrectOrNotSetFormatterFunction do
+    defexception [:message]
+  end
+end
+
+defimpl MetaLogger.Formatter, for: Any do
+  @moduledoc """
+  Default implementation for any struct which used to be derived.
+  """
+  defmacro __deriving__(module, _struct, options) do
+    formatter_func = fetch_formatter_function(options)
+
+    quote do
+      defimpl MetaLogger.Formatter, for: unquote(module) do
+        @replacement "[FILTERED]"
+
+        def format(struct) do
+          formatter_fn = Keyword.fetch!(unquote(options), :formatter_fn)
+
+          struct
+          |> Map.get(:payload)
+          |> safe_func_invoke()
+          |> filter()
+        end
+
+        defp filter(payload) when is_list(payload),
+          do: Enum.map(payload, &filter(&1))
+
+        defp filter(payload) when is_bitstring(payload) do
+          Keyword.get(unquote(options), :filter_patterns)
+          |> filter(payload)
+        end
+
+        defp filter(patterns, payload) do
+          Enum.reduce(patterns, payload, fn pattern, payload ->
+            String.replace(payload, ~r/#{pattern}/, @replacement)
+          end)
+        end
+
+        defp safe_func_invoke(args) do
+          unquote(Macro.escape(formatter_func)).(args)
+        rescue
+          FunctionClauseError ->
+            reraise MetaLogger.Formatter.IncorrectPayload,
+                    [
```

Comment:
Do we need the brackets here?

Reply:
yep, because it's not `raise`, but `reraise`, which saves the previous stack trace (and it's credo recommendation to use it :) )